### PR TITLE
implement transpose

### PR DESCRIPTION
**Pull request by tatusamel**
## Context
I am implementing standalone shim versions of all the necessary operators for running the Whisper Model libtorch-free, thereby eliminating the dependency on ATen. This PR is related with `transpose.int` operator.
[Enabling TorchNative Standalone on Whisper](https://docs.google.com/document/d/1BOdW0MA6n8couHc24s32np4qRkED1F3QZteWgNitaUY/edit?usp=sharing)

## What does this PR do?
In this PR I implemented [transpose](https://www.internalfb.com/code/fbsource/[f82dc142901a]/fbcode/caffe2/aten/src/ATen/native/TensorShape.cpp?lines=3756) operator.

`transpose` operator is used both in cpu and cuda so I had to implement:
- `aoti_standalone_cpu_transpose_int()`
- `aoti_standalone_cuda_transpose_int()`

**Note**: we skipped some checks:
 - if the tensor is sparse
 - its layout checks
 - is_mkldnn() check.

## Testing

I've written 4 test in total. 3 for CPU and 1 for CUDA.

Test 1: TransposeOp:
    - primary test for transpose operation
    - on CPU
    - creates a standard 2x3 at::Tensor and a SlimTensor that is a non-owning view of the same memory
    - calls our shim function `aoti_torch_cpu_transpose_int()`
    - at the end checks these:
        - if the dimensions are correct (dim())
        - the sizes have been swapped from {2, 3} to {3, 2}
        - The strides have been swapped from {3, 1} to {1, 3}
        - The resulting tensor is correctly flagged as **not contiguous** 
        - Most importantly, it verifies that the `data_ptr()` is the same as the original, confirming it's a view and no data was copied
        
Test 2: TransposeNoOp
    - handles edge case where dims to transpose are the same
    - it checks if operation did nothing, as expected.
    - if the resulting SlimTensor has the exact same sizes and strides, and is_contiguous flag as the original tensor.

Test 3: TransposeOfTranspose
    - Transposing a tensor twice returns it to its original state
    - First it transposes the original tensor and gets an intermediate result `t1_result`. 
    - Second, it transposes the intermediate tensor (`t1_result`)
    - At the end it checks:
        - `t2_result` has metadata that perfectly matches the original tensor and is correctly flagged as is_contiguous again.  

Test 4: TransposeOpCUDA
    - verifies if transpose works on cuda device
    - it creates a at::Tensor and SimTensor view on the CUDA device
    - it calls our shim function
    - then performs the same metada checks as before tests.

You can run `./build/bin/test_aoti_standalone` to run the tests.

Output:

```
(pytorch-3.12) [serhatgundem@devvm2480.eag0 /data/users/serhatgundem/pytorch_fork (transpose_int)]$ ./build/bin/test_aoti_standalone
Only one CUDA device detected. Disabling MultiCUDA tests
Note: Google Test filter = *-*_MultiCUDA
[==========] Running 11 tests from 3 test suites.
[----------] Global test environment set-up.
[----------] 4 tests from SlimTensorTest
[ RUN      ] SlimTensorTest.ResizeOpShrinkCPU
[       OK ] SlimTensorTest.ResizeOpShrinkCPU (0 ms)
[ RUN      ] SlimTensorTest.ResizeOpGrowCPU
[       OK ] SlimTensorTest.ResizeOpGrowCPU (0 ms)
[ RUN      ] SlimTensorTest.ResizeOpToZeroCPU
[       OK ] SlimTensorTest.ResizeOpToZeroCPU (0 ms)
[ RUN      ] SlimTensorTest.ResizeOpGrowCUDA
[       OK ] SlimTensorTest.ResizeOpGrowCUDA (0 ms)
[----------] 4 tests from SlimTensorTest (1 ms total)

[----------] 3 tests from SlimTensorInternalTest
[ RUN      ] SlimTensorInternalTest.SetSizesContiguous
[       OK ] SlimTensorInternalTest.SetSizesContiguous (0 ms)
[ RUN      ] SlimTensorInternalTest.SetSizesAndStridesNonContiguous
[       OK ] SlimTensorInternalTest.SetSizesAndStridesNonContiguous (0 ms)
[ RUN      ] SlimTensorInternalTest.EmptyTensorRestride
[       OK ] SlimTensorInternalTest.EmptyTensorRestride (0 ms)
[----------] 3 tests from SlimTensorInternalTest (0 ms total)

[----------] 4 tests from TransposeTest
[ RUN      ] TransposeTest.TransposeOp
[       OK ] TransposeTest.TransposeOp (0 ms)
[ RUN      ] TransposeTest.TransposeNoOp
[       OK ] TransposeTest.TransposeNoOp (0 ms)
[ RUN      ] TransposeTest.TransposeOfTranspose
[       OK ] TransposeTest.TransposeOfTranspose (0 ms)
[ RUN      ] TransposeTest.TransposeOpCuda
[       OK ] TransposeTest.TransposeOpCuda (175 ms)
[----------] 4 tests from TransposeTest (175 ms total)

[----------] Global test environment tear-down
[==========] 11 tests from 3 test suites ran. (176 ms total)
[  PASSED  ] 11 tests.
```

### CODE REVIEW

File: torch/standalone/transpose_int_template.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+#include <torch/csrc/inductor/aoti_standalone/c/shim.h>
+#include <torch/csrc/inductor/aoti_standalone/utils.h>
+#include <algorithm>
+#include <vector>
+
+namespace torch::standalone {
+
+template <typename T>
+inline T transpose_template(const T& self, int64_t dim0, int64_t dim1) {
```
Comment (yushangdi) [inline torch/standalone/transpose_int_template.h:11]: nit: following the naming convention in resize, maybe we should  call it `_transpose`?

File: torch/standalone/transpose_int_template.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  int64_t ndim = self.dim();
+  dim0 = torch::standalone::maybe_wrap_dim(dim0, ndim);
+  dim1 = torch::standalone::maybe_wrap_dim(dim1, ndim);
+
+  // if dimensions are the same, return a copy that is an alias.
+  // TODO: check if true
```
Comment (yushangdi) [inline torch/standalone/transpose_int_template.h:17]: what's the TODO here?

Comment (tatusamel) [reply torch/standalone/transpose_int_template.h:17]: oh I put it for myself to handle the case where the passed dimensions are equal. I mean to include it in the unit test. I forgot to remove it. Thanks a lot.

File: torch/standalone/transpose_int_template.h (hunk 1 above)
Comment (tatusamel) [reply torch/standalone/transpose_int_template.h:11]: you're totally right. I'll change it now. Thank you

File: torch/standalone/transpose_int_template.h [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  int64_t ndim = self.dim();
+  dim0 = torch::standalone::maybe_wrap_dim(dim0, ndim);
+  dim1 = torch::standalone::maybe_wrap_dim(dim1, ndim);
+
+  // if dimensions are the same, return a copy that is an alias.
+  if (dim0 == dim1) {
+    return self;
+  }
+
+  std::vector<int64_t> new_sizes(self.sizes().begin(), self.sizes().end());
```
[the remark this answers is not in the document]
Comment (tatusamel) [reply torch/standalone/transpose_int_template.h:20]: oh didn't know that. Thank you

**Review by desertfire — APPROVED** [review]
